### PR TITLE
feat: pixi rattler build

### PR DESCRIPTION
**Pull request by nichmor**
## Overview:
In this PR we add rattler-build as a new backend. This means that user can build a `recipe.yaml` if it is next to a`pixi.toml` or in `recipe` folder that is nearby a `pixi.toml`.
It is somehow similar with current existing `conda-build` backend, but instead of calling it directly and having implementation done on pixi side, we lift all work on pixi-build-backend side using jsonrpc ( similar with pixi-build-python ) .

* Added `--ignore-recipe` option for `pixi build` to ignore recipe.yaml if it's present and build pixi.toml
This is done by adding  a `DiscoveryConfig`for protocol builder which is responsible for controlling what protocol we want to discover.
This means that , by using `pixi build --ignore-recipe` we can ignore recipe.yaml build just pixi.toml. This means that it is also possible to configure the behaviour, and for example ignore both `pixi.toml` and `recipe.yaml` and build just meta.yaml.
* We now have a `BaseProtocol` trait. This means that both pixi, and rattler-build will just implement it. It also allow us to hide all jsonrpc details inside of `BaseProtocol`. In future, conda-build should also be implemented as a `BaseProtocol`.

For the next PR's:
* Add support for `recipe.yaml` directly in [dependencies] and make pixi build it.

### CODE REVIEW

File: crates/pixi_build_frontend/Cargo.toml [hunk 1]
```diff
@@ -10,6 +10,7 @@ repository.workspace = true
 version = "0.1.0"
 
 [dependencies]
+async-trait = { workspace = true }
```
Comment (tdejager) [inline crates/pixi_build_frontend/Cargo.toml:13]: We might be able to use rust async trait support for this?

File: crates/pixi_build_frontend/src/protocol_builder.rs [hunk 2]
```diff
@@ -5,10 +5,32 @@ use rattler_conda_types::ChannelConfig;
 use crate::{
     conda_build_protocol, pixi_protocol,
     protocol::{DiscoveryError, FinishError},
+    rattler_build_protocol,
     tool::ToolCache,
     BackendOverride, BuildFrontendError, Protocol,
 };
 
+#[derive(Debug)]
+pub struct DiscoveryConfig {
```
Comment (tdejager) [inline crates/pixi_build_frontend/src/protocol_builder.rs:14]: Needs docs, and not sure I like the name `DiscoveryConfig` because it seems to enable Protocols. I know this influences discovery but feels like implementation detail. Maybe something like `EnabledProtocolImplementations` or something shorter?

File: crates/pixi_build_frontend/src/protocol_builder.rs [hunk 3]
```diff
@@ -17,6 +39,10 @@ pub(crate) enum ProtocolBuilder {
     /// A directory containing a `meta.yaml` that can be interpreted by
     /// conda-build.
     CondaBuild(conda_build_protocol::ProtocolBuilder),
+
+    /// A directory containing a `recipe.yaml` that can be interpreted by
```
Comment (tdejager) [inline crates/pixi_build_frontend/src/protocol_builder.rs:43]:
```suggestion
    /// A directory containing a `recipe.yaml` that can be built with
```

File: crates/pixi_build_frontend/src/protocols/mod.rs [hunk 4]
```diff
@@ -1,4 +1,381 @@
...
+        }
+    }
+}
+
+/// Protocol trait that is responsible to setup and communicate with the backend.
+/// This allow us to hide the jsonrpc communication hidden in this protocol.
+/// This protocol is generic over the manifest what are passed to the build backends.
+/// This means that, for rattler-build, the manifest is a recipe.yaml file,
+/// and for pixi it's a pixi.toml/pyproject.toml file.
+#[async_trait::async_trait]
```
Comment (tdejager) [inline crates/pixi_build_frontend/src/protocols/mod.rs:104]: We could try using rust async fn in trait support?

File: crates/pixi_build_frontend/src/protocols/pixi/mod.rs [hunk 5]
```diff
@@ -140,6 +140,11 @@ impl ProtocolBuilder {
         )
         .await?)
     }
+
+    /// Returns the manifest of pixi
```
Comment (tdejager) [inline crates/pixi_build_frontend/src/protocols/pixi/mod.rs:144]:
```suggestion
    /// Returns the pixi manifest
```

File: crates/pixi_build_frontend/src/protocols/mod.rs (hunk 4 above)
Comment (tdejager) [reply crates/pixi_build_frontend/src/protocols/mod.rs:104]: Hmm actually, I rather that this a struct and named something like `JsonRPCBuildProtocol`. Essentially, pixi-build-backend and rattler-build-backend are user of this protocol and these things:

```rust
    /// Returns a unique identifier for the backend.
    fn backend_identifier(&self) -> &str;

    /// Returns the client used to communicate with the backend.
    fn client(&self) -> &Client;

    /// Returns a unique identifier for the build.
    fn build_id(&self) -> usize;

    /// Returns the handle to the stderr of the backend process.
    fn stderr(&self) -> Option<Arc<Mutex<Lines<BufReader<ChildStderr>>>>>;
```

with the exclusion of `stderr` can be passed in. This feels like it could all just be passed into the struct I've mentioned above. It feels like this is emulating inheritance where it need not be. `stderr` might be a method that can be implemented on the struct directly.

Basically, its all the same protocol but just using different data. A trait, in my opinion would make more sense, if we would abstract over different kind of protocols, like XMLRPC vs JsonRPC.

**Review by tdejager — CHANGES_REQUESTED** [review]

**Review by tdejager — APPROVED** [review]
Would also be great to add tests for the specific errors that you expect, so that we can make sure they keep looking nice.

**Review by tdejager — APPROVED** [review]
I think the main thing I would like to see more clear is that `RattlerBuild`, `PixiBuild` etc are users of the `BuildProtocol` and are not re-implementing it.

File: crates/pixi_build_frontend/src/build_frontend.rs [hunk 6]
```diff
@@ -88,14 +92,22 @@ impl BuildFrontend {
         }
     }
 
+    /// Sets the discovery config.
+    pub fn with_discovery_config(self, discovery_config: EnabledProtocols) -> Self {
```
Comment (tdejager) [inline crates/pixi_build_frontend/src/build_frontend.rs:96]: This is still called discovery_config.

File: crates/pixi_build_frontend/src/build_frontend.rs [hunk 7]
```diff
@@ -88,14 +92,22 @@ impl BuildFrontend {
...
+
     /// Constructs a new [`Protocol`] for the given request. This object can be
     /// used to build the package.
     pub async fn setup_protocol(
         &self,
         request: SetupRequest,
     ) -> Result<Protocol, BuildFrontendError> {
         // Determine the build protocol to use for the source directory.
-        let protocol = ProtocolBuilder::discover(&request.source_dir)?
+        let protocol = ProtocolBuilder::discover(&request.source_dir, &self.discovery_config)?
```
Comment (tdejager) [inline crates/pixi_build_frontend/src/build_frontend.rs:110]: Here, the same.

File: crates/pixi_build_frontend/src/protocols/mod.rs [hunk 8]
```diff
@@ -1,4 +1,403 @@
...
+            e => Self::JsonRpc(backend_identifier, e),
+        }
+    }
+}
+
+/// Protocol trait that is responsible to setup and communicate with the backend.
+/// This allow us to hide the jsonrpc communication hidden in this protocol.
+/// This protocol is generic over the manifest what are passed to the build backends.
+/// This means that, for rattler-build, the manifest is a recipe.yaml file,
+/// and for pixi it's a pixi.toml/pyproject.toml file.
```
Comment (tdejager) [inline crates/pixi_build_frontend/src/protocols/mod.rs:100]:
```suggestion
/// and for pixi it's a pixi.toml or a pyproject.toml file.
```

File: crates/pixi_build_frontend/src/protocols/mod.rs [hunk 9]
```diff
@@ -1,4 +1,403 @@
...
+    manifest_path: PathBuf,
+
+    _backend_capabilities: BackendCapabilities,
+
+    stderr: Option<Arc<Mutex<Lines<BufReader<ChildStderr>>>>>,
+}
+
+impl JsonRPCBuildProtocol {
+    /// Return the manifests that are used as input globs.
+    // fn manifests(&self) -> Vec<String>;
```
Comment (tdejager) [inline crates/pixi_build_frontend/src/protocols/mod.rs:122]:
```suggestion
```

File: crates/pixi_build_frontend/src/protocols/mod.rs [hunk 10]
```diff
@@ -1,4 +1,403 @@
...
+        cache_dir: Option<PathBuf>,
+        tool: Tool,
+    ) -> Result<Self, InitializeError> {
+        match tool.try_into_executable() {
+            Ok(tool) => {
+                // Spawn the tool and capture stdin/stdout.
+                let mut process = tokio::process::Command::from(tool.command())
+                    .stdout(std::process::Stdio::piped())
+                    .stdin(std::process::Stdio::piped())
+                    .stderr(std::process::Stdio::piped()) // TODO: Capture this?
```
Comment (tdejager) [inline crates/pixi_build_frontend/src/protocols/mod.rs:161]:
```suggestion
                    .stderr(std::process::Stdio::piped())
```

File: crates/pixi_build_frontend/src/protocols/stderr.rs [hunk 11]
```diff
@@ -8,7 +8,7 @@ use tokio::{
 
 /// Stderr sink that captures the stderr output of the backend
 /// but does not do anything with it.
-pub(crate) async fn stderr_null(
+pub async fn stderr_null(
```
Comment (tdejager) [inline crates/pixi_build_frontend/src/protocols/stderr.rs:11]: Does the visibility need a change here?

File: examples/boltons/pixi.toml [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+platforms = ["osx-arm64"]
+version = "0.1.0"
+
+[tasks]
+
+[dependencies]
+boltons = { path = "." }
+
+[host-dependencies]
+hatchling = "*"
```
Comment (tdejager) [inline examples/boltons/pixi.toml:15]:
```suggestion
hatchling = "*"
```

maybe pin this :)

File: src/cli/build.rs [hunk 13]
```diff
@@ -38,6 +38,11 @@ pub struct Args {
     /// Use system backend installed tool
     #[arg(long, action = ArgAction::SetTrue)]
     pub with_system: bool,
+
+    /// If recipe.yaml is present in the source directory, ignore it
```
Comment (tdejager) [inline src/cli/build.rs:42]:
```suggestion
    /// If a recipe.yaml is present in the source directory, ignore it
```

File: tests/integration_python/build/test_build.py [hunk 14]
```diff
@@ -44,3 +45,92 @@ def test_build_conda_package(pixi: Path, tmp_path: Path) -> None:
...
+
+    # Create a new project
+    verify_cli_command([pixi, "init", tmp_path])
+
+    parsed_manifest = tomllib.loads(manifest_path.read_text())
+    parsed_manifest["host-dependencies"] = {"hatchling": "*"}
+    parsed_manifest["build"] = {
+        "build-backend": "pixi-build-rattler-build",
+        "channels": [
+            # until rattler-build backend is merged in pixi build backend,
```
Comment (tdejager) [inline tests/integration_python/build/test_build.py:61]: I think this should be merged actually :)

https://prefix.dev/channels/pixi-build-backends/packages/pixi-build-rattler-build

Also graf cannot ever run on any other platform then arm-osx, so cant really merge then :)

**Review by tdejager — CHANGES_REQUESTED** [review]

File: tests/integration_python/build/test_build.py (hunk 14 above)
Comment (nichmor) [reply tests/integration_python/build/test_build.py]: changed!

**Review by tdejager — APPROVED** [review]
